File: src/main/java/de/unruh/onthefly/psi/impl/SimplePsiImplUtil.java
package de.unruh.onthefly.psi.impl;

import de.unruh.onthefly.Expression;
import de.unruh.onthefly.FakeComputationInlay;
import de.unruh.onthefly.PsiExpression;
import de.unruh.onthefly.inlays.Inlay;
import de.unruh.onthefly.psi.*;

import java.util.List;

public class SimplePsiImplUtil {
  public static String getVariable(SimpleAssignment element) {
    return element.getNode().findChildByType(SimpleTypes.IDENTIFIER).getText();
  }

    public static Expression getExpression(SimplePrefixExpression element) {
    return new Expression.App(
            element.getPrefixOp().getText(),
            element.getAtom().getExpression()
    );
  }

  public static Expression getExpression(SimpleInfixExpression element) {
    List<SimpleAtom> atoms = element.getAtomList();
    assert atoms.size() == 2;
    return new Expression.App(
            element.getInfixOp().getText(),
            atoms.get(0).getExpression(), atoms.get(1).getExpression());
  }

  public static Expression getExpression(SimpleExpression element) {
    PsiExpression e = element.getInfixExpression();
    if (e != null) return e.getExpression();
    e = element.getPrefixExpression();
    assert e != null;
    return e.getExpression();
  }

  public static Expression getExpression(SimpleAtom element) {
    PsiExpression e = element.getVariable();
    if (e != null) return e.getExpression();
    e = element.getNumeral();
    assert e != null;
    return e.getExpression();
  }

  public static Expression getExpression(SimpleNumeral element) {
    return new Expression.Num(element.getText());
  }

  public static Expression getExpression(SimpleVariable element) {
    return new Expression.Var(element.getText());
  }

  public static Inlay getInlay(SimpleAssignment element) {
    return new FakeComputationInlay(element.getVariable(), element.getExpression().getExpression());
  }
}

File: src/main/java/de/unruh/onthefly/unsorted/SimplePsiImplUtil.scala
package de.unruh.onthefly.unsorted

import de.unruh.onthefly.unsorted.Expression
import de.unruh.onthefly.unsorted.PsiExpression
import de.unruh.onthefly.inlays.Inlay
import de.unruh.onthefly.unsorted.psi._
import de.unruh.onthefly.unsorted.FakeComputationInlay
import java.util
import java.util.List

import com.intellij.psi.PsiElement

import scala.runtime.Nothing$


object SimplePsiImplUtil {

  def getExpression(element: SimplePrefixExpression) = new Expression.App(element.getPrefixOp.getText, element.getAtom.getExpression)

  def getExpression(element: SimpleInfixExpression): Expression = {
    val atoms = element.getAtomList
    assert(atoms.size == 2)
    new Expression.App(element.getInfixOp.getText, atoms.get(0).getExpression, atoms.get(1).getExpression)
  }

  def getExpression(element: SimpleExpression): Expression = {
    var e = element.getInfixExpression : PsiExpression
    if (e != null) return e.getExpression
    e = element.getPrefixExpression
    assert(e != null)
    e.getExpression
  }

  def getExpression(element: SimpleAtom): Expression = {
    var e = element.getVariable : PsiExpression
    if (e != null) return e.getExpression
    e = element.getNumeral
    assert(e != null)
    e.getExpression
  }

  def getExpression(element: SimpleNumeral): Expression.Num = Expression.Num(element.getText)

  def getExpression(element: SimpleVariable): Expression.Var = Expression.Var(element.getText)

}